src/CQRS/LeanCode.CQRS.AspNetCore/Local/MiddlewareBasedLocalCommandExecutor.cs

```diff
@@ -0,0 +1,81 @@
+using LeanCode.Contracts;
+using LeanCode.CQRS.Execution;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace LeanCode.CQRS.AspNetCore.Local;
+
+public class MiddlewareBasedLocalCommandExecutor : ILocalCommandExecutor
+{
+    private readonly IServiceProvider serviceProvider;
+
+    private readonly RequestDelegate pipeline;
+
+    public MiddlewareBasedLocalCommandExecutor(
+        IServiceProvider serviceProvider,
+        Action<ICQRSApplicationBuilder> configure
+    )
+    {
+        this.serviceProvider = serviceProvider;
+
+        var app = new CQRSApplicationBuilder(new ApplicationBuilder(serviceProvider));
+        configure(app);
+        app.Run(LocalCommandExecutor.HandleAsync);
+        pipeline = app.Build();
+    }
+
+    public async Task<CommandResult> RunAsync<T>(
+        HttpContext context,
+        T command,
+        CancellationToken cancellationToken = default
+    )
+        where T : ICommand
+    {
+        await using var scope = serviceProvider.CreateAsyncScope();
+
+        var localContext = new LocalHttpContext(context, scope.ServiceProvider, cancellationToken);
+
+        localContext.SetCQRSRequestPayload(command);
```

Comment:
Yeah, I think that is true.

I'm starting to wonder if we should split the "metadata" about the request from the "metadata" about execution here somehow, so that we don't have to resolve `HandlerType` too early.

Reply:
What do you mean by 'request metadata`? Command type? Endpoint routing?

If we assume one type can only be at most one of command/query we can query `CQRSObjectsRegistrationSource` to get `CQRSObjectMetadata`. Then we have all the information to finalize the pipeline.